Make heroku build idempotent

dhollinger opened the pull request.

dhollinger commented on cmd/goplugin-heroku/resource/build.go:
```diff
@@ -99,6 +100,18 @@ func (*BuildHandler) Update(ctx context.Context, externalID string, desiredState
+		return desiredState, nil
+	} else {
+		hclog.Default().Debug("Continuing update", "desiredState", desiredState)
+	}
```
Query the Build and check the buildpack name with the returned URL

thallgren commented on cmd/goplugin-heroku/resource/build.go:
```diff
@@ -253,3 +266,25 @@ func uploadSource(filePath, httpMethod, httpUrl string) error {
 	return nil
 }
+
+func buildpackCheck(desiredState Build, currentState heroku.Build) bool {
```
This function returns `true` when at least one buildpacks exists, when the first build pack either contains or doesn't contain the same name in the first desired versus current build pack, and when the all remaining buildpacks have the same characteristics as the first one. That's an algorithm that is hard to understand . A comment describing the rationale would be helpful.

thallgren commented on cmd/goplugin-heroku/resource/build.go:
```diff
@@ -253,3 +266,25 @@ func uploadSource(filePath, httpMethod, httpUrl string) error {
+	var result bool
+
+	for i, packs := range desiredState.Buildpacks {
+		if strings.Contains(currentState.Buildpacks[i].URL, *packs.Name) {
```
Is `len(desiredState.Buildpacks) <= len(currentState.Buildpacks)` guaranteed? If not, then this might fail with an out of bounds panic. Also, if `len(desiredState.Buildpacks) < len(currentState.Buildpacks)` can be true, then I guess the excess entries in `currentState` should affect the outcome somehow.

dhollinger replied:
Good point on the first part. There is a chance that `len(desiredState.Buildpacks) > len(currentState.Buildpacks)`, at which point we'd want an automatic false generated to indicate that the desired state has more buildpacks than exist on the app requiring an update.

As for the second part, the tricky part with buildpacks is that an app always requires a buildpack, but the user doesn't have to explicitly name one. If one isn't named, heroku will attempt to figure out what is needed based on what the build contains. So `currentState` will always have a buildpack if a build exists, but the `desiredState` doesn't have to have one defined.

That said, it'd probably be a good idea to check `len(desiredState.Buildpacks)` and if it is `= 0`, then automatically ignore the Buildpacks field. The other option would be to require that a user define a buildpack to make lyra's life easier

thallgren replied:
When `len(desiredState.Buildpacks) > len(currentState.Buildpacks)`, this line will still panic.

dhollinger replied:
It shouldn't ever reach that line if `len(desiredState.Buildpacks > len(currentState.Buildpacks)`. 

See https://github.com/lyraproj/lyra/pull/342/files#diff-de40d7ec6f34bc2a1d332f56d2c58ddaR328

dhollinger commented:
@thallgren Updated PR.

Added comment detailing the purpose of what `buildpackCheck` does and noted that the Heroku Builds API always keeps the list of buildpacks in the same order they were uploaded allowing for a direct comparison of both slices based on one slice's element id.

Added a new function that re-validates the desiredState and currentState of the build to ensure that we aren't skipping an update when there needs to be one or updating if there doesn't need to be one. This same method will compare the lengths of the `desiredState.Buildpacks` and `currentState.Buildpacks` slices. 

* If the `desiredState.Buildpacks == 0`, then always add a `true` to the status slice 
* If the `desiredState.Buildpacks` slice is larger than the `currentState.Buildpacks` slice, then always add a `false` to the status slice.
* All other occasions, run the `buildpackCheck` function.

thallgren commented on cmd/goplugin-heroku/resource/build.go:
```diff
@@ -253,3 +264,79 @@ func uploadSource(filePath, httpMethod, httpUrl string) error {
+	if len(status) < 2 {
+		return status[0]
+	} else {
+		for p := 0; p < len(status); p++ {
```
This seems odd to me. If `status[0]` is `false`, and if there is only one status, then this function returns `false` (line 291). However, if `len(status) == 2` and both `status[0]` and `status[1]` are `false`, then this function returns `true`. Why is that?

dhollinger commented:
@thallgren Any updates on this PR?
